server/api/lib/mavat/index.js

```diff
@@ -39,22 +39,99 @@ const init = () =>
 		})();
 	});
 
-const downloadFile = (url, file, entityDocId, entityDocNumber) => {
+const downloadChallengedFile = (url, file, options) => {
 	return new Promise((resolve) => {
-		http.get(url, (response) => {
-			response.pipe(file);
-			file.on('finish', async function () {
-				await file.close();
-				resolve(true);
-			});
+		options = options || {};
+
+		http.get(url, options, (response) => {
+			if (response.statusCode !== 200) {
+				Log.info(`downloadChallengedFile failed with status ${response.statusCode} for url ${url}`);
+				resolve(false);
+			} else {
+				const contentType = response.headers['content-type'] || '';
+
+				// if content-type is text/html this isn't the file we wish to download but one
+				// of the challenge stages
+				if (contentType.startsWith('text/html')) {
+					// if we didn't get a cookie yet this is the first part of the challenge -
+					// the page source contains the javascript code we need to run and challenge
+					// paramters for the calculation
+					if (!('set-cookie' in response.headers)) {
+						// download the entire response so we can solve the challenge
+						let responseData = '';
```

Comment:
Maybe we could extract to an external function, code too long- line 61 to 80

Reply:
is the general logic clear in your opinion?

i feel like the fact that the whole challenge logic is written in order here makes it clearer than if each part was a split-off function and you had to jump up and down the source to read the logic